### ISSUE TITLE
Computing a (mixture of hyperparameters) posterior predictive -- with a custom method. 

### ISSUE THREAD

**Issue by vr308** [cornellius-gp/gpytorch#1896]
I've trained a GP and I have a bunch of samples from the hyperparameter posterior -- (say I have this from some other program, not pyro) but I have a trace of samples for each hyperparameter and I can access it like in a dictionary. All I am trying to do at test time is **compute a predictive posterior per hyper sample in this trace** . The method below was written with the intention of returning a list of posterior predictives (one at each hyperparam sample). 

But it seems like this doesn't work as intended and the pred distributions seems to have a non-psd covariance matrix **for all but the very first sample.**  -- seems like I am missing some detail about updating the model to the fresh hyperparameters during eval time in loop? 

```
def mixture_posterior_predictive(model, test_x, trace_hyper):
     
      ''' Returns the posterior predictive multivariate normal '''
     
      model.eval()
      model.likelihood.eval()
      
      # Make predictions by feeding model through likelihood
      list_of_y_pred_dists = []
      
      for i in range(len(trace_hyper)):
          
         # extract a hyper sample
         hyper_sample = trace_hyper[i]

         # Update model with the particular sample
         model.likelihood.noise_covar.noise = hyper_sample['sig_n']**2
         model.base_covar_module.outputscale = hyper_sample['sig_f']**2
         model.base_covar_module.base_kernel.lengthscale = hyper_sample['ls']
       
         with torch.no_grad():
              # compute the posterior predictive for that sample
              pred = model.likelihood(model(test_x))
              try:  # check if the covariance is positive semi-definite 
                    chol = torch.linalg.cholesky(pred.covariance_matrix + torch.eye(len(test_x))*1e-5)
                    list_of_y_pred_dists.append(pred)
              except RuntimeError:
                   print('Not psd for sample ' + str(i))
     
      return list_of_y_pred_dists
```

**Comment by wjmaddox** [comment]
Try placing a `model.train()` statement before each posterior call to clear any training caches. Otherwise, you probably want to doublecheck that the length and output-scales are actually positive.

**Comment by jacobrgardner** [comment]
To clarify, you'll need to switch the model to train mode as @wjmaddox suggests and then back to eval model:
```
model.train()
model.eval()
```
That will almost surely fix your problem I think.

**Comment by vr308** [comment]
Just confirming -> it is `model.train()` before the overwriting hypers block and then switch to `model.eval()` just before the posterior predictive call? like this...

```
def mixture_posterior_predictive(model, test_x, trace_hyper):
     
      ''' Returns the mixture posterior predictive multivariate normal '''
     
      # Make predictions by feeding model through likelihood
     
      list_of_y_pred_dists = []
      
      for i in range(len(trace_hyper)):
          
         hyper_sample = trace_hyper[i]
         
         ## Training mode for overwriting hypers 
         model.train()
         model.likelihood.train()
         
         model.likelihood.noise_covar.noise = hyper_sample['sig_n']**2
         model.base_covar_module.outputscale = hyper_sample['sig_f']**2
         model.base_covar_module.base_kernel.lengthscale = hyper_sample['ls']
         
         with torch.no_grad():
             
              ## Testing mode for computing the posterior predictive 
              model.eval()
              model.likelihood.eval()
              
              pred = model.likelihood(model(test_x))
              
              try:
                    chol = torch.linalg.cholesky(pred.covariance_matrix + torch.eye(len(test_x))*1e-5)
                    list_of_y_pred_dists.append(pred)
              except RuntimeError:
                   print('Not psd for sample ' + str(i))
     
      return list_of_y_pred_dists
```

**Comment by wjmaddox** [comment]
Yes, that should be fine.
